### ISSUE TITLE
artificial time drift in live mode for native HLS (0.2.18)

### ISSUE THREAD

**Issue by jfairley** [clappr/clappr#668]
This is a weird one. I've tracked down the source, but I'm not sure of the correct way to fix it.

My scenario is that I have a live HLS feed. I seek back so that I'm in DVR mode. Every 10 seconds, the current time reported from `PLAYER_TIMEUPDATE` jumps 10 seconds. Watching the video proves that the video is not actually jumping 10 seconds forward, but the current time value artificially gets 10 seconds larger at regular intervals.

After some investigations, I see that this is definitely tied to when new playable fragments load. IE: when the duration gets 10 seconds longer, the current time is also 10 seconds longer.

I've traced this as far as I feel comfortable to the following call stack:

1. [when dispatching the timeupdate event](https://github.com/clappr/clappr/blob/0.2.18/src/playbacks/hls/hls.js#L80)
2. [`this.getCurrentTime()` returns the wrong value](https://github.com/clappr/clappr/blob/0.2.18/src/playbacks/hls/hls.js#L45)
3. [which seems to have bad input coming from here](https://github.com/clappr/clappr/blob/0.2.18/src/playbacks/hls/hls.js#L19)
4. [and I don't know why `this.getDuration()` never changes](https://github.com/clappr/clappr/blob/0.2.18/src/playbacks/html5_video/html5_video.js#L200)

I added a `console.log` into `getPlayableStartTime()` to show the weirdness:

![screen shot 2015-11-13 at 4 29 09 pm](https://cloud.githubusercontent.com/assets/992483/11160311/5aa41302-8a25-11e5-90a5-979d16630f08.png)

What you see is `this.getDuration()` (1st number) staying the same while the actual duration (2nd number) steadily increases in 10-second chunks (as expected). This causes `getPlayableStartTime()` to be an ever-growing negative number, in turn causing "current time" as dispatched in the event to grow artificially.

- Any idea what the fix is here?
- Where does `this.el.duration` come from / who is responsible for updating it? Can it be ignored?

**Comment by dviramontes** [comment]
+1 on this issue

**Comment by tjenkinson** [comment]
@jfairley thanks for reporting this.

`getPlayableStartTime()` is meant to return the time into the video element where content actually starts. If you look at a live stream on http://dailymotion.github.io/hls.js/demo/ you'll see the way it works is that the total duration of the video element continues to increase throughout the stream, but the video data is removed from the element as the stream goes on. At least this is what I understood happens. Maybe I'm wrong?

Therefore the duration reported directly from the video element will always be increasing and is not the duration of the actual video content in the video element. `playableRegionDuration` should be the duration of the available video content.

From your log though it looks like the video element duration is not constantly increasing which is what I'd have expected.

**Comment by tjenkinson** [comment]
`this.getDuration()` just returns the duration reported from the video element, which I would have expected to be constantly increasing, and if it was then I don't think there'd be this issue.

Maybe the duration the plain video element shows on it's standard control bar on the hlsjs demo page is different to a value it reports on `.duration` . I just assumed it would be the same.

**Comment by tjenkinson** [comment]
Yep just checked in chrome and the duration that is reported on the control bar is not the same as the duration it will report from it's api. Looks like the video api actually reports the amount of video content it's actually holding. This will be the root of the issue.

**Comment by jfairley** [comment]
So another interesting thing... If I'm close enough to the live end of the stream (but still not live), I see something similar but different.

Every 10 seconds, duration jumps instantaneously (because it came from an event), then `this.el.duration` races to catch up. You see an upward slide in the 1st number.

The effect on my seek bar is rather strange. Every 10 seconds, the play head jumps forward 10 seconds, then slides backwards to where it is supposed to be.

![screen shot 2015-11-13 at 5 16 35 pm](https://cloud.githubusercontent.com/assets/992483/11160705/dab919f2-8a2a-11e5-8db5-22fdbb00c785.png)

**Comment by tjenkinson** [comment]
@jfairley is this in chrome 46.0.2490.86 ?

**Comment by jfairley** [comment]
I see it in
- Chrome 48.0.2560.0 dev (64-bit)
- Chrome 48.0.2563.0 canary (64-bit)
- Opera 32.0.1948.69

I actually don't have vanilla Chrome installed.

**Comment by tjenkinson** [comment]
Looks like maybe the `seekable` property should be used instead of `duration`, but the value that is returning doesn't seem to make sense at the moment, not sure if it's a bug with hlsjs or not at the moment.

**Comment by jfairley** [comment]
BTW, in the meantime, I've found that bypassing the getters in clappr seems to work for me. Hopefully this is just a temporary work-around, but it's definitely interesting that it works as well as it does...

```javascript
    onPlayerTimeUpdate(progress, duration) {
        var videoElement = document.querySelector('video');
        progress = videoElement.currentTime;
        duration = videoElement.duration;
        TimeActions.timeUpdate(progress, duration);
    }
```

I also had to bypass the built-in seek function, since it uses the suspect value for duration. In it, I found that `seekSeconds` works really really well. I know there's another discussion ongoing about that.

```javascript
    onSeek(seekTo) {
        window.player.core.mediaControl.container.playback.seekSeconds(seekTo);
        //window.player.seek(100 * seekTo / TimeStore.getDuration());
    }
```

**Comment by tjenkinson** [comment]
Interesting.

Well it's very likely this is wrong

```js
 timeUpdated() {
    if (this.dvrEnabled) {
      this.trigger(Events.PLAYBACK_TIMEUPDATE, this.dvrInUse ? this.getCurrentTime() : this.playableRegionDuration, this.playableRegionDuration, this.name)
    } else {
      super.timeUpdated()
    }
  }
```
as the duration doesn't always map to the current time if it's a sliding window.

**Comment by tjenkinson** [comment]
Just to confirm if you are playing a stream with a sliding window you would expect the current time to increase normally, and then jump back every time a new chunk loads right?

So for example:
0:21
0:22
0:23
0:24
0:25
0:21
0:22
...

If each chunk duration was 5 seconds.

You wouldn't be expecting the current time to keep increasing even when content from the beginning is removed?

**Comment by jfairley** [comment]
First, I'm not sure what it means to have a stream that is a sliding window. Does that mean that media from the beginning of the stream is eventually unavailable? If that is so, I have no opinion on what current time looks like in that use case as I can't say I'm familiar such a UX.

While my streams are live streams, the user can always seek back to the beginning. It's more like DVR functionality from your cable provider. You can record a TV show and either watch it as it happens or catch it later. But either way, you can always view the entirety of what has already happened. As duration grows, I expect current time to continue growing also. I expect current time to never be less than a previous value.

**Comment by FDiskas** [comment]
:+1:

**Comment by tjenkinson** [comment]
Yes it means that content from the start is removed as more content is appended to the end as the stream is essentially infinitely long.

Ok yes I am testing with a stream like yours now and get the issue. If you change
```js
this.trigger(Events.PLAYBACK_TIMEUPDATE, this.dvrInUse ? this.getCurrentTime() : this.playableRegionDuration, this.playableRegionDuration, this.name)
```
to
```js
this.trigger(Events.PLAYBACK_TIMEUPDATE, this.getCurrentTime(), this.playableRegionDuration, this.name)
```

That should fix it. Here's a build with that change: https://gist.github.com/tjenkinson/afd696c05b3758c2ead1

@jfairley please can you try it?

Looking more into why that was added now.

The other changes you made will work for your case, because the duration is always the duration of all the content as nothing is ever deleted, but probably won't for cases with a sliding window as that is not the case, but I'm looking into that as well.

**Comment by tjenkinson** [comment]
Ok so looking into it further `.duration` on the video element should never be used because the value of this appears to be the time from the beginning to the time that the last segment in the video element finishes, and HLSJS decides when it wants to download segments and which ones.

E.g. your stream has been going for an hour and you rewind back to 10 minutes, HLSJS will start filling the gap from the 10 minute point onwards, and will stop appending video to the end given you're not watching it, meaning `.duration` will stop increasing.

`data.details.totalduration` reported from HLSJS should be used in its place instead, which is currently what `playableRegionDuration` is.

@mangui are my assumptions correct about the following?

if `video.duration` is 70, but HLSJS's `totalDuration` is reported as 100, would setting `currentTime` on the video element to 90 be valid, and cause the chunks from 90 onwards then to start buffering?

**Comment by FDiskas** [comment]
Are we talking about this bug? https://youtu.be/4Kg54ztFTcY

**Comment by tjenkinson** [comment]
I have started a branch which I'm putting the fixes in: https://github.com/clappr/clappr/tree/feat-hls-dvr-fixes

@FDiskas I don't have that issue any more with my latest changes. Please can you try again with the latest build from that branch (https://gist.github.com/tjenkinson/4b8ebf73b2d04d7c6be9)?

**Comment by FDiskas** [comment]
OK,
so I build this from source of your branch with `npm install` and `npm run build`
to many errors. Please merge the master branch first

**Comment by tjenkinson** [comment]
@FDiskas I can't merge it to the master branch because it's not ready for that yet. You should be able to use the files in dist folder of the master branch but then use this file (https://gist.githubusercontent.com/tjenkinson/4b8ebf73b2d04d7c6be9/raw/11a8a4a59d975cb3da31dc9155f8a9cc931ee610/clappr.js) instead of the clappr.js in that folder.

**Comment by FDiskas** [comment]
But you can merge master branch in to yours branch

**Comment by tjenkinson** [comment]
My branch is only 1 commit ahead of master at the moment with the extra commit being the changes I made.

**Comment by FDiskas** [comment]
https://youtu.be/640uxazmkDo

**Comment by FDiskas** [comment]
And for some time I cant click on seek-bar - unresponsive.

**Comment by tjenkinson** [comment]
Yes clicking on the seek bar won't work until it switches to dvr mode. The seek bar is lagging behind a bit when a new chunk comes in for some reason which it shouldn't be.

So far all I've been working on is trying to get the times that are reported from the player correct, which I should be able to sort once I get an update on https://github.com/dailymotion/hls.js/issues/65

One of the changes I have made results in it jumping around before DVR becomes enabled so that needs sorting next.

**Comment by tjenkinson** [comment]
I do not have that lagging issue locally either @FDiskas please can you provide the stream url you're using?

**Comment by FDiskas** [comment]
Private url can't share

**Comment by tjenkinson** [comment]
Right the seek bar jumping issue is down to the current hack in `getPlayableStartTime()`, so this is essentially blocked now waiting on https://github.com/dailymotion/hls.js/issues/65

**Comment by tjenkinson** [comment]
@FDiskas @jfairley the latest build of my branch should work for you (https://gist.githubusercontent.com/tjenkinson/19c0fbb1f8fa0c2c7eff/raw/3daa243b5e6182719067b3feeb9f1e98960220f6/clappr.js)

It currently will not support sliding window streams. Let me know if there are issues.

**Comment by jfairley** [comment]
@tjenkinson I didn't get to try out your changes, but @lkkadiri was telling me that it was looking pretty good.

You're right about not wanting to rely only on duration from the video tag. Although, our scenario improved greatly, we did start seeing funkiness when we started seeking back and pausing during long streams. Let us know if we can test out any further changes.

**Comment by FDiskas** [comment]
@tjenkinson looks like almost perfect. https://www.youtube.com/watch?v=m30ln9yAk9Y

**Comment by tjenkinson** [comment]
@FDiskas an issue I see there looks like it might be related to the dvr controls plugin. I think this issue is finished now please can you create new ones for any new issues, unless they are related to this one. Thanks

@jfairley Here's the build of the latest version that's in the pull request waiting to be merge in: https://gist.githubusercontent.com/tjenkinson/ccbafc1d08403cd6a424/raw/159557280cd78a40acf79f277c243a594e59e8ee/clappr.js
If you still have some seek issues please can you create another issue for those?

**Comment by jfairley** [comment]
:+1: Thanks a bunch!
